compass/step.py

```diff
@@ -210,6 +236,88 @@ def __init__(self, test_case, name, subdir=None, cores=1, min_cores=1,
+        ``configure()`` or ``run()`` for the test case that this step belongs
+        to.
+
+        Parameters
+        ----------
+        cpus_per_task : int, optional
+            the number of cores per task the step would ideally use.  If
+            fewer cores per node are available on the system, the step will
+            run on all available cores as long as this is not below
+            ``min_cpus_per_task``
+
+        min_cpus_per_task : int, optional
+            the number of cores per task the step requires.  If the system
+            has fewer than this number of cores per node, the step will fail
+
+        ntasks : int, optional
+            the number of tasks the step would ideally use.  If too few
+            cores are available on the system to accommodate the number of
+            tasks and the number of cores per task, the step will run on
+            fewer tasks as long as as this is not below ``min_tasks``
+
+        min_tasks : int, optional
+            the number of tasks the step requires.  If the system has too
+            few cores to accommodate the number of tasks and cores per task,
+            the step will fail
+
+        openmp_threads : int, optional
+            the number of OpenMP threads to use
+
+        max_memory : int, optional
+            the amount of memory that the step is allowed to use in MB.
+            This is currently just a placeholder for later use with task
+            parallelism
+        """
+        if cpus_per_task is not None:
+            self.cpus_per_task = cpus_per_task
+        if min_cpus_per_task is not None:
+            self.min_cpus_per_task = min_cpus_per_task
+        if ntasks is not None:
+            self.ntasks = ntasks
+        if min_tasks is not None:
+            self.min_tasks = min_tasks
+        if openmp_threads is not None:
+            self.openmp_threads = openmp_threads
+        if max_memory is not None:
+            self.max_memory = max_memory
+
+    def constrain_resources(self, available_cores):
+        """
+        Constrain ``cpus_per_task`` and ``ntasks`` based on the number of
+        cores available to this step
+
+        Parameters
+        ----------
+        available_cores : int
+            The total number of cores available to the step
+        """
+        if self.ntasks == 1:
+            # just one task so only need to worry about cpus_per_task
+            self.cpus_per_task = min(self.cpus_per_task, available_cores)
+            cores = self.cpus_per_task
+        elif self.cpus_per_task == self.min_cpus_per_task:
+            available_tasks = available_cores // self.cpus_per_task
+            self.ntasks = min(self.ntasks, available_tasks)
+            cores = self.ntasks * self.cpus_per_task
+        else:
+            raise ValueError('Unsupported step configuration in which '
+                             'ntasks > 1 and  '
+                             'cpus_per_task != min_cpus_per_task')
```

Comment:
To preserve general sanity, I'm enforcing a rule that `cpus_per_task == min_cpus_per_task` if `ntasks > 1`.  If you're doing MPI jobs, you will always know that `cpus_per_task = min_cpus_per_task = openmp_thread`, since you're using `cpus_per_task` for OpenMP threading.  (This will be enforced in a future update.)  If you are running a multithreaded or multiprocessing job on a single node, you will always have `ntasks = 1` and the number of threads or processes will be controlled by `cpus_per_task` and `min_cpus_per_task`.

Reply:
Should `cpus_per_task = min_cpus_per_task = openmp_thread` be `cpus_per_task = min_cpus_per_task * openmp_thread`?

Reply:
No, if you're running with threads, the number of OpenMP threads is the same thing as the number of CPUs per MPI task.  This is just by definition.

In addition to that, what I'm also saying here is that things get too confusing if `cpus_per_task` is different from `min_cpus_per_task`.  That is, it gets very confusing if a developer allows for some number of MPI tasks in a range between `min_tasks` and `ntasks`, and at the same time they also allow for a range of threads between `min_cpus_per_task` and `cpus_per_task`.  I don't want to deal with that complicated (maybe even impossible) optimization problem.  So I suggest we also require that the number of OpenMP threads is always fixed for tasks with multiple MPI tasks, so that `cpus_per_task = openmp_threads` and `min_cpus_per_task = cpus_per_task`.

Does that make things any clearer?

Reply:
Yes, that makes sense, thanks.